File: nodejs-io/src/contract/terminal_contract.test.ts
import { describe, it, expect } from "vitest";
import { create } from "@bufbuild/protobuf";
import {
  OpenTerminalRequestSchema,
  TerminalOutputSchema,
  type OpenTerminalRequest,
  type TerminalOutput,
} from "../gen/agent/v1/terminal_pb";
import { loadGolden } from "./golden_loader.js";

describe("Terminal Contract (Open)", () => {
  it("OpenTerminalRequest maps golden fields onto camelCase logical fields", () => {
    const golden = loadGolden("terminal_open.req.json") as {
      pipeline_id: string;
      container_id: string;
      mode: string;
      actor: string;
    };

    const req: OpenTerminalRequest = create(OpenTerminalRequestSchema, {
      pipelineId: golden.pipeline_id,
      containerId: golden.container_id,
      mode: golden.mode as OpenTerminalRequest["mode"],
      actor: golden.actor,
    });

    expect(req.pipelineId).toBe("p-001");
    expect(req.containerId).toBe("ctr-abc");
    expect(req.mode).toBe("TERMINAL_MODE_READ_ONLY");
    expect(req.actor).toBe("user@example.com");
  });

  it("TerminalOutput maps golden fields onto camelCase logical fields", () => {
    const golden = loadGolden("terminal_open.resp.json") as {
      stream: string;
      data: string; // base64-encoded bytes
      is_done: boolean;
    };

    const data = new Uint8Array(Buffer.from(golden.data, "base64"));

    const resp: TerminalOutput = create(TerminalOutputSchema, {
      stream: golden.stream,
      data,
      isDone: golden.is_done,
    });

    expect(resp.stream).toBe("stdout");
    expect(Buffer.from(resp.data).toString()).toBe("hello\n");
    expect(resp.isDone).toBe(false);
  });
});

File: test/contract/terminal_contract_test.go
package contract

import (
	"testing"

	agentv1 "openforge/gen/go/agent/v1"
	"github.com/stretchr/testify/require"
)

func TestContract_TerminalOpen_RequestShape(t *testing.T) {
	golden := loadGolden(t, "golden/terminal_open.req.json")
	req := &agentv1.OpenTerminalRequest{}
	assertRoundTrip(t, golden, req)

	require.Equal(t, "p-001", req.GetPipelineId())
	require.Equal(t, "ctr-abc", req.GetContainerId())
	require.Equal(t, agentv1.TerminalMode_TERMINAL_MODE_READ_ONLY, req.GetMode())
	require.Equal(t, "user@example.com", req.GetActor())
}

func TestContract_TerminalOpen_ResponseShape(t *testing.T) {
	golden := loadGolden(t, "golden/terminal_open.resp.json")
	resp := &agentv1.TerminalOutput{}
	assertRoundTrip(t, golden, resp)

	require.Equal(t, "stdout", resp.GetStream())
	require.Equal(t, []byte("hello\n"), resp.GetData())
	require.False(t, resp.GetIsDone())
}
